spring-integration-core/src/main/java/org/springframework/integration/dsl/EndpointSpec.java

```diff
@@ -141,4 +143,24 @@ protected void assertHandler() {
 		Assert.state(this.handler != null, "'this.handler' must not be null.");
 	}
 
+	protected MessageChannel obtainInputChannelFromFlow(IntegrationFlow subFlow) {
+		return obtainInputChannelFromFlow(subFlow, true);
+	}
+
+	protected MessageChannel obtainInputChannelFromFlow(IntegrationFlow subFlow, boolean evaluateInternalBuilder) {
```

Comment:
Why false on routers and true everywhere else (including recipient list router)? Perhaps javadocs on these methods would help.

Reply:
Yeah... That's hard to explain so far. See `route(S routerSpec, Consumer<S> routerConfigurer)`.
I think the main reason to do that to be able to add beans into the `BeanFactory` in the proper order. So, since sub-flows (somehow) in router depend on their root router component, I register it first and only after that I build sub-flows from their builder.

Looks like in other place we don't have such a strong dependency, so it's fine to register sub-flows before their main component.

I thought briefly to change such a logic, but decided to do that in the separate JIRA (if that is possible at all) and already without backporting.

Right,  I'll add JavaDocs to them to explain things a bit.

Thanks

P.S. I won't mind if you (or someone else) will take a look to this subject independently. Another pair of eyes sometimes gives a good unexpected solution.